packages/progress-bar/src/ProgressBar.tsx

```diff
@@ -0,0 +1,98 @@
+import { Box, Heading } from "@kaizen/component-library"
+import React from "react"
+import classnames from "classnames"
+import styles from "./ProgressBar.scss"
+
+type Props = {
+  value: Value
+  variant: Variant
+  mood: Mood
+  subtext?: string
+  labelExtraContent?: React.ReactNode
+}
+
+type Value =
+  | { kind: "percentage"; value: number }
+  | { kind: "fraction"; value: number; max: number }
```

Comment:
This is a "discriminated union type", where the two options are discriminated by the "kind".

FWIW, this is what it would look like in Elm (it's the same idea):

```
type Value =
  Percentage Int
  | Fraction { value: Int, max: Int }
```

Reply:
Personally, i'm not a fan of the nested prop. An object is likely going to be re-instantiated by the parent on a re-render, which means the progress bar is going to re-render each time. Now, that isn't the biggest deal in the world, it's not likely to cause performance issues or anything, but it could be a dangerous precedent to set for the design system. This component could get more complex over time, or other components could be built with this API as prior art. I believe the components from the design system should be as stable as possible, and if that comes at the expense of a more difficult to read type error, so be it.

But I think I have a seperate concern with the union itself. Max isn't a prop that can't exist when things are a percentage, rather, it _should_ be 100. If the progress bar were to have a default value of 100 for its max, then max can be an optional prop. With this, any value passed can be treated as a percentage and max can be ignored by the consumer in the default case. In the case of fractions, the max can be specified as anything else. The resulting progress is the same calculation and the rendered result is the same.

Where this breaks down is the label. But I don't think the type _should_ make the label change. The label seems like a perfectly reasonable thing for a consumer to want to control at render time. If we allow that to exist they can render:

* `${value}%`
* `${value} percent`
* `${value}/${max}`
* `${value} with any extra content you like`

The consumer can now also choose _not_ to render a label. We'd also no longer need to care about the `extraContent`.

Reply:
I think I like the simpler approach as well. I don't think it's unreasonable for this to only accept a percentage value rather than offering the ability to calculate the percentage from a fraction.

Moving the responsibility of the label onto the consumer seems like the most future proof solution, however I do wonder about the accessibility impact of no label - mainly because the colour contrast might be hard for some to see, and the label provides a way to see that information with sufficient contrast. It also means that consumers will need to support RTL themselves, which will probably be patchy.

Tough one, but I'm still leaning towards giving the consumer the responsibility of the label and crossing that bridge when we get there.

Reply:
I've changed the API so it's just `?label`. Let me know if that's not what you meant (or feel free to add a commit with exactly what you were thinking).